MinioAdmin: allow specifying policies as dict besides file

Alveel opened the pull request:
It would be nice to be able to add/update IAM/service account policies through Python dict objects as well as what's currently available through JSON files.

This PR tries to accomplish this.

Feedback/questions welcome.

Depends on https://github.com/minio/minio-py/pull/1483

darix commented:
i wonder if it makes sense to refactor out something like

```
def _load_policy(policy_file, policy):
        if policy and policy_file:
            raise ValueError(
                "only one of policy or policy_file may be specified")
        if not policy or not policy_file:
            raise ValueError("either policy or policy_file must be specified")
        body = ''
        if policy:
            body = json.dumps(policy)
        elif policy_file:
            with open(policy_file, encoding='utf-8') as file:
                body = file.read()
       return body
```

give in how many places we basically have the same code now.

Alveel commented:
Sounds fairly sensible actually. I'll wait for feedback from a member and see what they think.

balamurugana commented on minio/minioadmin.py:
```diff
@@ -452,16 +452,30 @@ def group_list(self) -> str:
-                body=file.read().encode(),
-            )
-            return response.data.decode()
+        if policy and policy_file:
```
Use XOR wherever applicable
```py
if (policy_file is not None) ^ (policy is not None):
    body = json.dumps(policy).encode() if policy is not None else pathlib.Path(policy_file).read_bytes()
else:
    raise ValueError("only one of policy_file or policy must be provided")
```

Alveel commented on minio/minioadmin.py:
```diff
@@ -111,6 +112,18 @@ def _safe_str(value: Any) -> str:
+def _load_policy(policy_file: str | None = None,
+                 policy: dict | None = None) -> str:
+    """Load a policy from file or dictionary."""
+    if (policy_file is None) ^ (policy is None):
```
I flipped the conditions didn't I. I flipped my tests too. Will fix.

harshavardhana commented:
> I'd love to hear some feedback on the current state.

Please fix the CI tests

Alveel commented:
Fixed. Moved the `.open()` for `Path` to its own line so I can put the mypy comment on the same line while keeping it within 80 characters to keep autopep8 happy.

balamurugana commented on minio/minioadmin.py:
```diff
@@ -763,7 +782,7 @@ def add_service_account(self,
             raise ValueError("both access key and secret key must be provided")
```
Add xor check of `policy_file` and `policy` here and fail early.

balamurugana commented on minio/minioadmin.py:
```diff
@@ -111,6 +112,19 @@ def _safe_str(value: Any) -> str:
 
 
+def _load_policy(policy_file: str | None = None,
+                 policy: dict | None = None) -> dict:
```
1. This generic function does not work for `add_service_account()` and `update_service_account()`. In the APIs, `policy_file is not None and policy is not None` is an error case, other combinations are not.
2. Add checks for `policy_file` and `policy` in respective APIs. This function is not needed.

Alveel replied:
I'm not exactly sure if I follow entirely. From how I view this:

- Specifying `policy_file` and `policy` is an error case
- Specifying only `policy_file` is OK
- Specifying only `policy` is OK
- Specifying neither for `policy_add()` is an error case, this is OK for `add_service_account()` and `update_service_account()`.

Am I missing something?

I did found a bug that I just now fixed, but otherwise this works in my testing. Can you elaborate on why this does not work?

Do you prefer to have this code completely separated (and duplicated) across `policy_add()`, `add_service_account()` and `update_service_account()` over a separate function like this?

Thank you!

balamurugana replied:
@Alveel As I mentioned in previous comments, This function is not needed. Refer the diff for correct implementation.

balamurugana reviewed (changes requested):
Below is the exact diff, I am referring
```diff
diff --git a/minio/minioadmin.py b/minio/minioadmin.py
index d62ac21..0fd2979 100644
--- a/minio/minioadmin.py
+++ b/minio/minioadmin.py
@@ -452,16 +452,24 @@ class MinioAdmin:
         response = self._url_open("GET", _COMMAND.LIST_GROUPS)
         return response.data.decode()
 
-    def policy_add(self, policy_name: str, policy_file: str) -> str:
+    def policy_add(self,
+                   policy_name: str,
+                   policy_file: str | None = None,
+                   policy: bytes | None = None) -> str:
         """Add new policy."""
-        with open(policy_file, encoding='utf-8') as file:
-            response = self._url_open(
-                "PUT",
-                _COMMAND.ADD_CANNED_POLICY,
-                query_params={"name": policy_name},
-                body=file.read().encode(),
-            )
-            return response.data.decode()
+        if not (policy_file is not None) ^ (policy is not None):
+            raise ValueError("either policy_file or policy must be provided")
+        body = policy
+        if policy_file:
+            with open(policy_file, encoding='utf-8') as file:
+                body = file.read().encode()
+        response = self._url_open(
+            "PUT",
+            _COMMAND.ADD_CANNED_POLICY,
+            query_params={"name": policy_name},
+            body=body,
+        )
+        return response.data.decode()
 
     def policy_remove(self, policy_name: str) -> str:
         """Remove policy."""
@@ -754,6 +762,7 @@ class MinioAdmin:
                             name: str | None = None,
                             description: str | None = None,
                             policy_file: str | None = None,
+                            policy: dict | None = None,
                             expiration: str | None = None,
                             status: str | None = None) -> str:
         """
@@ -763,7 +772,9 @@ class MinioAdmin:
             raise ValueError("both access key and secret key must be provided")
         if access_key == "" or secret_key == "":
             raise ValueError("access key or secret key must not be empty")
-        data = {
+        if policy_file is not None and policy is not None:
+            raise ValueError("either policy_file or policy must be provided")
+        data: dict[str, Any] = {
             "status": "enabled",
             "accessKey": access_key,
             "secretKey": secret_key,
@@ -775,6 +786,8 @@ class MinioAdmin:
         if policy_file:
             with open(policy_file, encoding="utf-8") as file:
                 data["policy"] = json.load(file)
+        if policy:
+            data["policy"] = policy
         if expiration:
             data["expiration"] = expiration
         if status:
@@ -798,15 +811,19 @@ class MinioAdmin:
                                name: str | None = None,
                                description: str | None = None,
                                policy_file: str | None = None,
+                               policy: dict | None = None,
                                expiration: str | None = None,
                                status: str | None = None) -> str:
         """Update an existing service account"""
-        args = [secret_key, name, description, policy_file, expiration, status]
+        args = [secret_key, name, description,
+                policy_file, policy, expiration, status]
         if not any(arg for arg in args):
             raise ValueError("at least one of secret_key, name, description, "
                              "policy_file, expiration or status must be "
                              "specified")
-        data = {}
+        if policy_file is not None and policy is not None:
+            raise ValueError("either policy_file or policy must be provided")
+        data: dict[str, Any] = {}
         if secret_key:
             data["newSecretKey"] = secret_key
         if name:
@@ -816,6 +833,8 @@ class MinioAdmin:
         if policy_file:
             with open(policy_file, encoding="utf-8") as file:
                 data["newPolicy"] = json.load(file)
+        if policy:
+            data["newPolicy"] = policy
         if expiration:
             data["newExpiration"] = expiration
         if status:
```

Alveel commented on minio/minioadmin.py:
```diff
@@ -452,16 +466,20 @@ def group_list(self) -> str:
+    def policy_add(self,
+                   policy_name: str,
+                   policy_file: str | None = None,
+                   policy: dict | None = None) -> str:
```
That's why I also changed this line:

https://github.com/Alveel/minio-py/blob/de89e7301767dba8bc8713144a5270c06d8e13c1/minio/minioadmin.py#L480

to encode the json string into bytes, so in the end bytes is always sent! :)

However, I can understand if the API should remain the same across SDK's. Is this what you meant?

balamurugana replied:
Yes. See the diff I pasted.

Alveel commented:
I think that's it? I still see "1 requested change" but I don't see where or what it is.

Alveel commented:
Squashed commits, should be all clean now. Can you please take a look?

balamurugana commented on minio/minioadmin.py:
```diff
@@ -452,16 +453,26 @@ def group_list(self) -> str:
-                body=file.read().encode(),
-            )
-            return response.data.decode()
+        if policy_file:
+            with Path(policy_file).open(encoding='utf-8') as file:
+                body = file.read().encode()
+        elif policy:
+            body = json.dumps(policy).encode()
+        else:
+            raise ValueError("either policy or policy_file must be specified")
+
+        response = self._url_open(
+            "PUT",
+            _COMMAND.ADD_CANNED_POLICY,
+            query_params={"name": policy_name},
+            body=body,
+        )
+        return response.data.decode()
```
```suggestion
        if not (policy_file is not None) ^ (policy is not None):
            raise ValueError("either policy_file or policy must be provided")
        body = policy
        if policy_file:
            with open(policy_file, encoding='utf-8') as file:
                body = file.read().encode()
        response = self._url_open(
            "PUT",
            _COMMAND.ADD_CANNED_POLICY,
            query_params={"name": policy_name},
            body=body,
        )
        return response.data.decode()
```

balamurugana commented on minio/minioadmin.py:
```diff
@@ -763,7 +775,9 @@ def add_service_account(self,
         if access_key == "" or secret_key == "":
             raise ValueError("access key or secret key must not be empty")
-        data = {
+        if policy_file and policy:
+            raise ValueError("specify either policy_file or policy, not both")
```
```suggestion
        if policy_file is not None and policy is not None:
            raise ValueError("either policy_file or policy must be provided")
```

balamurugana commented on minio/minioadmin.py:
```diff
@@ -773,8 +787,10 @@ def add_service_account(self,
             data["description"] = description
         if policy_file:
-            with open(policy_file, encoding="utf-8") as file:
+            with Path(policy_file).open(encoding="utf-8") as file:
```
```suggestion
            with open(policy_file, encoding="utf-8") as file:
```

balamurugana commented on minio/minioadmin.py:
```diff
@@ -797,25 +813,31 @@ def update_service_account(self,
-        data = {}
+                             "policy_file, policy, expiration or status must "
+                             "be specified")
+        if policy_file and policy:
+            raise ValueError("specify either policy_file or policy, not both")
```
```suggestion
        if policy_file is not None and policy is not None:
            raise ValueError("either policy_file or policy must be provided")
```

balamurugana commented on minio/minioadmin.py:
```diff
@@ -797,25 +813,31 @@ def update_service_account(self,
             data["newDescription"] = description
         if policy_file:
-            with open(policy_file, encoding="utf-8") as file:
+            with Path(policy_file).open(encoding="utf-8") as file:
```
```suggestion
            with open(policy_file, encoding="utf-8") as file:
```

balamurugana commented on minio/minioadmin.py:
```diff
@@ -24,6 +24,7 @@
 import os
 from datetime import timedelta
 from enum import Enum, unique
+from pathlib import Path
```
```suggestion
```

balamurugana reviewed (dismissed):
Depends on https://github.com/minio/minio-py/pull/1483

balamurugana commented on minio/minioadmin.py:
```diff
@@ -452,16 +452,24 @@ def group_list(self) -> str:
-            return response.data.decode()
+        if not (policy_file is not None) ^ (policy is not None):
+            raise ValueError("either policy_file or policy must be provided")
+        body = policy
```
```suggestion
        body = json.dumps(policy).encode() if policy else None
```
